Order product count api requests orders store

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Adds a new orders data store that makes use of the orders API -> https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-orders
For now this only allows for retrieving/search for orders and getting the orders count.

Partially addresses #32524 

### How to test the changes in this Pull Request:

1. Load this branch and create a couple of orders and go to **WooCommerce > Home**
2. Open your console and run `wp.data.select('wc/admin/orders').getOrders({ status: 'completed'}, undefined);`
3. It should return `undefined` at first
4. Run it again, assuming the request finished, and it should return all the completed orders. You can try the above again with other items in the query, like `_fields: ['id']` for example, which should only return the `ids`, any query param defined here should work: https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-orders
5. Now do the same for the `getOrdersTotalCount` function, by running: `wp.data.select('wc/admin/orders').getOrdersTotalCount({ status: 'completed'}, undefined);`, this might run the correct number right away if you recently ran `getOrders` with the same query params, otherwise it should return `undefined` initially.
6. If you have the [Redux Devtools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en) extension, you can look for the `wc/admin/orders` store and make sure the state is updated correctly and the actions are triggered correctly.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm nx changelog <project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Add initial Order redux store

louwie17 committed:
Add unit tests for orders store

botwoo commented:
:bar_chart: Test reports for d21cbce7de1374128f00ae073ec4cca1944bec2f `Add unit tests for orders store` have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/33063/api/)
- E2E Test Report (TBC)

Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them.

louwie17 committed:
Update fetch with headers to parse error

botwoo commented:
:bar_chart: Test reports for 9509e194a39462952e6df5d56ac7bad14118ac10 `Update fetch with headers to parse error` have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/33063/api/)
- E2E Test Report (TBC)

Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them.

joshuatf commented on packages/js/data/src/orders/action-types.ts:
```diff
@@ -0,0 +1,8 @@
+export enum TYPES {
+	SET_ORDER = 'SET_ORDER',
```
Same comment as the products data store PR about updating this to the new naming pattern.

joshuatf commented on packages/js/data/src/orders/constants.ts:
```diff
@@ -0,0 +1,3 @@
+export const STORE_NAME = 'wc/admin/orders';
+
+export const WC_ORDER_NAMESPACE = '/wc/v3/orders';
```
Super nitpicky, but maybe `WC_ORDERS_NAMESPACE`?  I noticed a few places where the pluralization doesn't match up.

joshuatf commented on packages/js/data/src/orders/resolvers.ts:
```diff
@@ -0,0 +1,76 @@
+	message: string;
+};
+
+function* request( query: Partial< OrderQuery > ) {
```
Same comment as other PR around this method.  Would be great to make this a bit more reusable if it's not a ton of effort.

joshuatf commented on packages/js/data/src/orders/selectors.ts:
```diff
@@ -0,0 +1,54 @@
+import { WPDataSelector, WPDataSelectors } from '../types';
+
+export const getOrders = createSelector(
+	( state: OrderState, query: OrderQuery, defaultValue = undefined ) => {
```
`OrdersState`, `OrdersQuery`?

joshuatf commented on packages/js/data/src/orders/types.ts:
```diff
@@ -0,0 +1,158 @@
+	meta_data: OrderMetaData[];
+};
+
+export type OrderStatuses =
```
It is possible to filter these and add/remove order statuses.  We might need to keep this more open-ended to continue supporting custom order statuses.

louwie17 replied:
The only way to really keep it open ended is using `string` instead, but that seems kinda odd here, or go against TS.
A user should be able to overwrite the `Order` type relatively easy by creating a new Order type:
``` 
type NewOrder = Order & {
  status: NewStatus | OrderStatuses;
}
```
I could provide a Status generic for Orders and default to the above, if you think that would be of benefit?

louwie17 replied:
I did the latter as well.

joshuatf commented on packages/js/data/src/orders/types.ts:
```diff
@@ -0,0 +1,158 @@
+export type PartialOrder = Partial< Order > & Pick< Order, 'id' >;
+
+export type OrderQuery = BaseQueryParams & {
+	status:
```
Same comment around statuses here.

joshuatf commented on packages/js/data/src/types/query-params.ts:
```diff
@@ -0,0 +1,16 @@
+export type BaseQueryParams = {
```
💯

joshuatf reviewed (commented):
Looking good!  Left some comments on this one; most are similar to the comments left in the sibling products PR.

louwie17 committed:
Update action names, and consolidate request function

louwie17 committed:
Add chnagelog

louwie17 commented:
@joshuatf thanks for the review, I addressed your PR feedback.
One thing to note, I didn't make the change in the items data store to use the `request` method in this PR as it's already being done as part of the `products` data store PR.

joshuatf commented on packages/js/data/src/orders/reducer.ts:
```diff
@@ -0,0 +1,101 @@
+					Record< number, PartialOrder >
+				>( ( result, order ) => {
+					ids.push( order.id );
+					result[ order.id ] = order;
```
When fetching a specific field while not including `id`, the last piece record's information is assigned to all rows presumably because this assigns to IDs.

<img width="739" alt="Screen Shot 2022-05-25 at 10 52 32 AM" src="https://user-images.githubusercontent.com/10561050/170293437-8316dd2d-943d-4cbe-9ba6-ce678ab62718.png">
<img width="573" alt="Screen Shot 2022-05-25 at 10 50 52 AM" src="https://user-images.githubusercontent.com/10561050/170293609-3db92048-0ca1-45e1-91c4-d0ef88dc71b2.png">

louwie17 replied:
The fix for this has been addressed in https://github.com/woocommerce/woocommerce/pull/33063/commits/153d49771742d272a733ac62d0eeb2c2846d869a

louwie17 committed:
Update resolver to always include id and update reducer to prevent data overwrite

louwie17 commented:
@joshuatf I updated this PR with the fix for the `_fields` query without `id`, if you have time to have another look :)

joshuatf reviewed (approved):
Solid work, @louwie17!  This is testing well and LGTM 🚢
